### ISSUE TITLE
New 5 digit nerd font glyphs breaks right prompt

### ISSUE THREAD

**Issue by rwmitchell** [romkatv/powerlevel10k#2289]
I have IP set to display on my second line right side prompt.  I have the glyph set using:

typeset -g POWERLEVEL9K_IP_VISUAL_IDENTIFIER_EXPANSION='󰀑'

which should be this:

<img width="175" alt="image" src="https://user-images.githubusercontent.com/15944841/235708972-6e14c44e-e7ea-4510-b7ac-09bdb5cf2c04.png">

however, it throws off the left side prompt, backing over the space after prompt_char.  Any editing/tabbing causes the cursor to jump back/forth by 1 char.  Unsetting the glyph fixes the issue, but then there is no glyph.

Previously I had used F510 without any issue, but it has been removed from the latest fonts.

Not to sound like a whiny idjit, P10K is amazing!  Been using it for several years!  Thanks!!

**Comment by rwmitchell** [comment]
You are probably right, but it does break on both iTerm2 and Apple Terminal.

**Comment by romkatv** [comment]
We can try diagnosing with a screenshot. Please run the following command and post a screenshot of the output.

```zsh
printf '\UF510\033[31mX\033[0m\nXX\n\UF0011\033[31mX\033[0m\n'
```

**Comment by rwmitchell** [comment]
iTerm2:
![image](https://user-images.githubusercontent.com/15944841/235726454-5f724ced-6646-4c7f-b96c-6ae5ac84dc23.png)

Apple Terminal:
![image](https://user-images.githubusercontent.com/15944841/235726632-8ed72cc1-4d48-4ba7-ab2d-1c54acc237aa.png)

Same configuration for both.  The current font doesn't have the original glyph.

I changed the glyph back to the orig value.  While the glyph is not the one I want, it does work in both terminals.

**Comment by romkatv** [comment]
Sorry, I don't understand your comment. Can you run the same command again and this time post a screenshot where I can see the command (your comment made me suspect that you've changed it; but since I don't understand your comment, I'm not sure whether you actually did). You don't need to do it in two terminals -- choose one that you actually care about.

**Comment by rwmitchell** [comment]
![image](https://user-images.githubusercontent.com/15944841/235730448-2d1bfb76-0a70-4bce-98c9-f9ef3a03e58f.png)


I had temporarily swapped the comments back to use the original glyph to verify it still worked, even though is not the one intended.   I'm still using an older font with both glyphs in MacVim - these are the lines I've been enabling/disabling:

![image](https://user-images.githubusercontent.com/15944841/235730502-436ad8ad-267b-4435-b29c-f2ae45ebad8c.png)




Note: I tried using '\UF0011' to set the glyph but I got the actual string in the prompt, not the glyph.  Not sure why, as that  syntax works elsewhere and I even verified I was typing it correctly by changing the battery indicator.

The first line is the old glyph, the second line is the new, the third line has it unset.

**Comment by romkatv** [comment]
The screenshot looks fine. I'm assuming it was made with the font that is actually causing an issue with your prompt.

Please set the parameter this way:

```zsh
typeset -g POWERLEVEL9K_IP_VISUAL_IDENTIFIER_EXPANSION=$'\UUF0011'
```

And post a screenshot showing the problem with your prompt. Also attach `.p10k.zsh` as a file (you'll probably need to rename it to `p10k.txt` to appease github).

**Comment by rwmitchell** [comment]
![image](https://user-images.githubusercontent.com/15944841/235734401-865b5206-df60-40d0-845f-2c718a19fd1e.png)


[p10k.txt](https://github.com/romkatv/powerlevel10k/files/11375777/p10k.txt)

**Comment by romkatv** [comment]
I've tried your config on Apple Terminal. It works fine after fixing a few things that were producing errors:

1. Quoted `p10k segment` arguments in `prompt_myosicon` and `prompt_codename`. This is necessary because I don't have `OS_CODENAME` defined on my machine.
2. Removed `POWERLEVEL9K_TIME_VISUAL_IDENTIFIER_EXPANSION` because I don't have `emoji-clock`. Note that this parameter makes your prompt slower.

No visual artifacts with this config. You can try to debug this on your side as follows.

1. First, try running `RPS1=$'\UF0011' zsh -f` and see if that works fine. It should work just as well as `RPS1=X zsh -f`. Check that right prompt is painted in the correct spot and the cursor is where it should be. Press Ctrl-R and verify that the cursor position does not change. If `\UF0011` causes trouble here, it would indicate a bug in the terminal.
2. If the first step doesn't reveal any problems, try to pinpoint the problem in your `~/.p10k.zsh`. Remove as much from it as possible to make the config minimal while still exhibiting the issue. Start by removing all prompt segments that aren't material to demonstrating the issue.

P.S.

Your transient prompt functions can be improved like this:

```zsh
function p10k-on-pre-prompt()  { p10k display '1|2/right/wifi|2/left_frame'=show '2/left/time|2/right/dir'=hide }
function p10k-on-post-prompt() { p10k display '1|2/right/wifi|2/left_frame'=hide '2/left/time|2/right/dir'=show }
```

This will remove left frame from past prompts.

**Comment by rwmitchell** [comment]
No, the RPS1 doesn't work:

![image](https://user-images.githubusercontent.com/15944841/235900681-3560b089-e457-48aa-9594-27db6b2f84b9.png)

![image](https://user-images.githubusercontent.com/15944841/235900761-0c3ed337-8cff-40f0-8c87-d2507705e497.png)

![image](https://user-images.githubusercontent.com/15944841/235900855-79e70a7c-e37a-46de-b7c5-96a01e387011.png)

In the third screenshot, I had just hit TAB and then line changed from showing 'it' to 'iit' and the space was removed.  I did this in both iTerm2 and Apple Terminal with the same results.

Could the issue be something else?

```DQ%locale
LANG="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_CTYPE="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_ALL=
```

**Comment by rwmitchell** [comment]
Thanks for your time and considerable patience!

**Comment by rwmitchell** [comment]
FOUND!

Apparently it is a bug in zsh and fixed in 5.8.1
zsh 5.8.0.2-dev (x86_64-apple-darwin20.6.0)  <-- fails
zsh 5.8.1 (x86_64-apple-darwin21.0)          <-- works

**Comment by rwmitchell** [comment]
I had initially compiled zsh from source, back when Mac's zsh was older and lacked stuff, limiting my happiness.

Monterey is now up to 5.8.1 and I suspect you're on Ventura with 5.9.

Thanks again for all your help!  It was your comment that it worked for you, caused me to continue looking elsewhere.

**Comment by h0adp0re** [comment]
@rwmitchell hint hint, wink wink: `brew install zsh`.

**Comment by romkatv** [comment]
> @rwmitchell hint hint, wink wink: `brew install zsh`.

Or you could just use `/bin/zsh`.

**Comment by romkatv** [comment]
> FOUND!
> 
> Apparently it is a bug in zsh and fixed in 5.8.1 zsh 5.8.0.2-dev (x86_64-apple-darwin20.6.0) <-- fails zsh 5.8.1 (x86_64-apple-darwin21.0) <-- works

I looked at the commits between zsh 5.8 and 5.9 and there is nothing that could've caused this bug. Moreover, I cannot reproduce this in 5.8. My guess is that when you compiled zsh, you didn't pass `--enable-unicode9` to `configure`. As a result, the following command likely prints "2" in your self-compiled zsh:

```zsh
print -r -- ${(m)#:-$'\UF0011'}
```

It's supposed to print "1".

**Comment by rwmitchell** [comment]
I don't use brew.  If the source isn't available and/or I can't get it to compile successfully, I don't use it.

As I said I initially compiled zsh several years ago when the MacOS version was behind and lacked some new feature that something else I was using needed. No doubt I didn't enable unicode9 as I didn't understand the importance of it and must not have been in the default configuration.  I generally look thru the list of options for fun things but otherwise leave it to the defaults.

MacOS has since caught up to the latest zsh (on Ventura and Monterey is now newer than the last time I recompiled(*), so there is no reason not to use the OS provided version, especially since it works better.

(*) I did update/recompile zsh a couple weeks ago.  It initially seemed to work, but then crashes, so I left the previous version in place.  I didn't look into further into why as there wasn't anything I needed and it was  -dev release on a older OS.

:-) I looked at the commits too, after I had initially posted that and didn't see anything either that looked like a suspicious culprit but didn't go any further.  Kudos for tracking it fully down!  You must really like puzzles!

It was only because of an errant change In OMZ that prompted to get the latest nerd fonts, where they have now removed the previously 'only marked as obsolete' glyphs.  I had previously updated many of the glyphs I used, except this one because it caused issues.  As the old one was still there, I reverted and only commented in .p10k why I didn't use it.

ps: on p10k, personally I'd prefer if the config default to leaving most things disabled and only enabling what you need.  Kinda like how OMZ requires selecting which plugins to use.  All the extra stuff really clutters the output from env and set.  But I also realize I'm probably alone in not using many of the newer tools.  The awesomeness of p10k certainly outweighs the effort to bring over my changes after a p10k update/configure.
